fendermint/vm/topdown/src/sync.rs

```diff
@@ -0,0 +1,273 @@
+    pub fn start(self) {
+        let config = self.config;
+        let provider = self.parent_view_provider;
+        let agent = self.agent;
+        let query = self.committed_state_query;
+
+        let mut interval = tokio::time::interval(Duration::from_secs(config.polling_interval_secs));
+
+        tokio::spawn(async move {
+            loop {
+                interval.tick().await;
+
+                if let Err(e) = sync_with_parent(&config, &agent, &provider, &query).await {
+                    tracing::error!("sync with parent encountered error: {e}");
+                }
+            }
+        });
+    }
+}
+
+/// Syncing with parent with the below steps:
+/// 1. Get the latest height in cache or latest height committed increment by 1 as the
+///    starting height
+/// 2. Get the latest chain head height deduct away N blocks as the ending height
+/// 3. Fetches the data between starting and ending height
+/// 4. Update the data into cache
+async fn sync_with_parent<T: ParentFinalityStateQuery + Send + Sync + 'static>(
+    config: &Config,
+    agent_proxy: &Arc<IPCProviderProxy>,
+    provider: &Arc<Toggle<CachedFinalityProvider<IPCProviderProxy>>>,
+    query: &Arc<T>,
+) -> anyhow::Result<()> {
+    let last_recorded_height = if let Some(h) = last_recorded_height(provider).await? {
+        h
+    } else {
+        // cannot get starting recorded height, we just wait for the next loop execution
+        return Ok(());
+    };
+
+    let parent_chain_head_height = agent_proxy
+        .get_chain_head_height()
+        .await
+        .context("cannot fetch parent chain head")?;
+    // sanity check
+    if parent_chain_head_height < config.chain_head_delay {
+        tracing::debug!("latest height not more than the chain head delay");
+        return Ok(());
+    }
+
+    let ending_height = parent_chain_head_height - config.chain_head_delay;
+
+    tracing::debug!(
+        "last recorded height: {}, parent chain head: {}, ending_height: {}",
+        last_recorded_height,
+        parent_chain_head_height,
+        ending_height
+    );
+
+    if last_recorded_height == ending_height {
+        tracing::debug!(
+            "the parent has yet to produce a new block, stops at height: {last_recorded_height}"
+        );
+        return Ok(());
+    }
+
+    // we are going backwards in terms of block height, the latest block height is lower
+    // than our previously fetched head. It could be a chain reorg. We clear all the cache
+    // in `provider` and start from scratch
+    if last_recorded_height > ending_height {
+        let finality = query_starting_finality(query, agent_proxy).await?;
+        atomically(|| provider.reset(finality.clone())).await;
+        return Ok(());
+    }
+
+    // we are adding 1 to the height because we are fetching block by block, we also configured
+    // the sequential cache to use increment == 1.
+    let starting_height = last_recorded_height + 1;
```

Comment:
I still think it would be a cheap and worthy step to start fetching from `last_recorded_height`, rather than one after, to be able to detect reorgs. 

The previous check `last_recorded_height > ending_height` established that the parent chain hasn't got shorter. If it did, the comment says it could be a reorg (or the parent node is re-syncing). But it could be longer and still have gone through a reorg:

```
B0 - B1 - B3 - B4 - B5
        \
          B3' - B4' - B5' - B6'
```

Say that `last_recorded_height` is 5 because we have fetched B5, and there was a reorg. Depending on when we poll, the following things could happen:
* we see `B4'` with height 4 and we clear the cache
* we see `B5'` with height 5 and we skip saying there are no new blocks
* we see `B6'` with height 6 and we fetch it and append it after our B5

It wouldn't be difficult to just fetch the block header at the `last_recorded_height` 5 and:
* if it doesn't exist any more because the chain got shorter, it's a reorg
* if it exists but differs, it's a reorg
* if it exists and matches, we can go and append new blocks

I know you commented earlier that reorgs are difficult to deal with but it seems perplexing to see the cache being reset on one signal of it happening but the obviously wrong thing being done on a different case, all just to save re-fetching one header.

Reply:
Created a more elaborate proposal in https://github.com/consensus-shipyard/fendermint/issues/314

Reply:
Let me also track it in the code, so it is clear. Thanks!